tests/data/test_dataprovider.py

```diff
@@ -154,6 +154,23 @@ def test_market(mocker, default_conf, markets):
     assert res is None
 
 
+def test_ticker(mocker, default_conf, tickers, markets):
+    api_mock = MagicMock()
+    api_mock.markets = markets
+    api_mock.fetch_ticker = MagicMock(side_effect=lambda x: tickers().get(x))
+    exchange = get_patched_exchange(mocker, default_conf, api_mock=api_mock)
```

Comment:
Usually, (unless we're testing the exchange class) - we should mock out the Exchange's methods - not the CCXT methods (which is what using api_mock does).

I think a better (or rather, more aligned method as to how mocking is handled in the rest of the tests would be:

```suggestion
     mocker.patch('freqtrade.exchange.Exchange.fetch_ticker', ticker)
    exchange = get_patched_exchange(mocker, default_conf, mock_markets=True)
```

Now i didn't test this (so typos may be included) - but i think this should work.

Reply:
`ticker` does not exist, we only have `tickers` in conftest... and the mocking function should issue same exception in the case of non-existing pair, as the Exchange method it mocks, to verify that the exception is probably handled by the higher-level dataprovider method...

Reply:
well ... no?
the fixture ticker exists - line 326 of conftest ... :) 

I think you need to be clear what you're trying to test - it's either `fetch_ticker()` behavior (we already have tests for this in `test_exchange.py`...) or it's the behavior of the new dataprovider method - which is mostly a "pass-through".

I don't mind having the test as is ... i was just proposing a different method since you asked.

One thing: i think the `MagicMock(side_effect=` part on the fetch_ticker mock is not needed - you can monkey-patch that function directly by assigning the lambda to it...

Reply:
thanks for advices, got it

I definitely want to test dp.ticker() here (handling of exceptions from the lower level, etc), and Exchange.fetch_ticker() was not touched and it's testing is out of scope...

pls review the updated test